### ISSUE TITLE
[Bug] Structured config for torch optimizer

### ISSUE THREAD

**Issue by Vozf** [facebookresearch/hydra#1011]
# 🐛 Bug
## Description
I am using structured config for instantiating torch optimizer
```
@dataclass
class AdamConfig(OptimizerConfig):
    _target_: str = "torch.optim.Adam"
    lr: float = 0.001
    betas: tuple = (0.9, 0.999)
    eps: float = 1e-08
    weight_decay: float = 0
```
Later I init the optimizer with 
```
optimizer = instantiate(self.cfg.optimizer, self.model.parameters())
```
After reading this config I get wrong type for `betas` which should be tuple(or at least list)
```
>>> type(self.cfg.optimizer.betas)
<class 'omegaconf.listconfig.ListConfig'>
```
This leads to major performance degradation(2.25 it\s for ListConfig vs 2.45 it\s for normal adam initialization on my task)
to recreate it you can avoid all the config stuff and just use 
```
from omegaconf import ListConfig

torch.optim.Adam(
            self.model.parameters(),
            lr=0.001,
            # betas=ListConfig([0.9, 0.999]),
            betas=[0.9, 0.999],
            eps=1e-08,
            weight_decay=0.0,
        )
```
With ListConfig optimizer degrades performance as I mentioned.
This way ListConfig shouldn't be used when initializing object

**Comment by omry** [comment]
Thanks for reporting, this is a current limitation of OmegaConf and it will probably be very hard - if not impossible - to solve.
In your use case, how big is your model (number of parameters)?
Your performance might improve even more if you use a primitive tuple and not a list (or any MutableSequence).
Can you check if it makes a difference?
```
betas=(0.9, 0.999)
```
1. Please file a bug against OmegaConf to improve tuple support, currently tuples are treated as lists. This is a big change and will take a long time to implement, and alone will not be enough to really solve the problem.
2. Please file a bug against PyTorch suggesting to covert an input beta MutableSequence to a primitive tuple.

```
from omegaconf import ListConfig, OmegaConf
isinstance((1,2), MutableSequence)
Out[9]: False
isinstance([1,2], MutableSequence)
Out[10]: True
isinstance(OmegaConf.create((1,2)), MutableSequence)
Out[11]: True
isinstance(tuple(OmegaConf.create((1,2))), MutableSequence)
Out[12]: False
```

cc @romesco 
cc @tkornuta-nvidia

**Comment by Vozf** [comment]
I've got around 1 mil parameters. Tuple advantage is very small if it isn't an experimental error(2.38it\s vs 2.39it\s).
Anyway it doesn't seem like a big issue compared to 10% degradation with ListConfig
It also seems like that should be described in the docs if you position hydra as experiment manager so that people are aware of possible caveats.

**Comment by omry** [comment]
The perf drop is indeed not huge, but is unexpected.

Are you running on a GPU? Ideally the betas would be moved to the GPU in such a case and this means they will no longer be in a ListConfig.

**Comment by Vozf** [comment]
Single Gpu. Not sure how they are handled. It seems like they are staying on cpu for the whole time

**Comment by omry** [comment]
I filed https://github.com/omry/omegaconf/issues/392 to keep track of the changes needed in OmegaConf.
as I said, that alone will not be enough.

**Comment by omry** [comment]
@Vozf , I need a minimal repro that demonstrates the performance regression.
Can you provide one?

**Comment by Vozf** [comment]
Unfortunately can't provide the exact sources.
I'm based on https://github.com/NathanUA/U-2-Net. So you should be able to reproduce it replacing Adam in u2net_train.py

**Comment by omry** [comment]
I don't need the exact sources, I need a minimal repro.

**Comment by Vozf** [comment]
Ok, I see. For a bug to be fixed you must provide a minimal repo. Unfortunately got no time to do this. Therefore closing

**Comment by omry** [comment]
It certainly help. otherwise I need to spend time trying to reproduce every single bug reported instead of making progress on things that would benefit everyone.

Since this is an important problem to keep on the radar I am going to keep this open with your permission.

**Comment by omry** [comment]
On second though, I will create a new bug report that reflect the problem better.
